sys/random: module to generate unique PRNGs seeds

PeterKietzmann opened the pull request:
This module is meant to provide unique seeds for the pseudo random number generator. Sources used for the seed generation (so far):
1. HWRNG - if no HWRNG available
2. Combination of ADC values and CPUID (LUID) - if no ADC available
3. Combination of CPUID (LUID) and xtimer

Generation of unique CPUIDs for boards that don't provide one natively is an other topic. In it's current state I use `FEATURES_REQUIRED = periph_hwrng` to check whether this module is available on a board. This obviously leads to warnings on boards not supporting a HWRNG. #5065 seems like a promising candidate for that problem.

haukepetersen commented on Makefile.dep:
```diff
@@ -589,6 +589,11 @@ ifneq (,$(filter random,$(USEMODULE)))
         USEMODULE += tinymt32
     endif
+    # sources for random seed
+    FEATURES_REQUIRED += periph_hwrng
```
now that I think about it, how about `FEATURES_OPTIONAL` as done with `periph_rtc` in the `default` example?

haukepetersen commented on boards/iotlab-m3/include/board.h:
```diff
@@ -82,6 +82,14 @@ extern "C" {
+ * @{
+ */
+#define RANDOM_SEED_ADC_LINE         (0)
+#define RANDOM_SEED_ADC_RES          ADC_RES_12BIT
```
These are the default values, right? So need to define them here...

PeterKietzmann replied:
I did it on purpose to demonstrate how/where to place these parameters, if needed. However, I moved  this config to stm32f0discovery

haukepetersen commented on sys/include/random.h:
```diff
@@ -43,7 +43,9 @@ extern "C" {
- * @param s seed for the PRNG
+ * @param s seed for the PRNG. If 0 (as in auto_init), a more or
+ *          less unique seed will be generated internally, dependent
+ *          on hardware ressources
```
This is not what is done anymore, right?

haukepetersen commented:
I still think I would like it better to put the seed code in its own module. Also I think it is quite important that we add some documentation to the seed header, that describes the quality of seed you can expect (and how it is influenced by available hardware units), and what you should use the seed value for (and what not).

kaspar030 commented:
Some thoughts:

- using xtimer_now() at a fixed point in the bootup process will lead to the same value on most MCUs unless sth unpredictable (timewise) happens before

- not all rngs are seeded by a 32bit value (luckily)

PeterKietzmann commented:
@haukepetersen can you give me a hint how to generate the module you proposed? I thought that I already created a module in the current way. @kaspar030 I'm using xtimer in case the seed is initialized later during runtime. Concerning the 32-bit what do you propose?

PeterKietzmann commented:
btw: thanks for the review @haukepetersen and @kaspar030

haukepetersen commented:
I think my comment about the module is outdated, so never mind. Although I think we can improve the naming of the module, how about simply `seed` with `seed_get_u32(void)`? 

Further we should detangle the dependencies in `Makefile.dep` a bit...

PeterKietzmann commented:
Just "seed" seems a bit too generic to me. How about `random_seed_get_u32()` instead of `random_prng_seed()`?

haukepetersen commented:
why too generic? Keeping it simple...

PeterKietzmann commented:
Simple "seed" is not so speaking imo [http://www.manythings.org/sentences/words/seed/1.html](http://www.manythings.org/sentences/words/seed/1.html). I don't care too much. Maybe a third opinion helps? 
De-tangling the dependencies in `Makefile.dep` seems reasonable but I still couldn't figure out how. E.g. checking for `FEATURES_PROVIDED` makes no sense as it isn't parsed at that point in time

PeterKietzmann commented:
@haukepetersen please have a look at the adapted `Makefile.dep`. The change can be found in a separate PR (#6852). We still need to agree on naming. @cgundogan can you state your opinion about the name of the random seed generator module?

miri64 commented:
@kaspar030 murdock seems to hang on this PR.

kaspar030 commented:
> @kaspar030 murdock seems to hang on this PR.

Fixed. Every now and then Murdock1 fills up its /tmp, making it stall on further PRs.

haukepetersen commented:
How do we proceed with this PR? Is the issue that came up in #6852 blocking this PR?

Except from that, I think I can live with the changes proposed here, although I have the feeling we the code can still be optimized (but let's leave that to follow-ups...).

PeterKietzmann commented:
@haukepetersen yes this PR was blocked by #6852. I just rebased but Murdock will not pass. Can think about it again this afternoon and I gladly accept different proposals regarding the Makefile issues.

PeterKietzmann commented:
@haukepetersen please have a look at this PR again. Murdock agrees now.
- for now I don't see an better solution for the Makefile.dep (compare  #6852)
- boards with unsatisfied features may throw a warning but build anyway
- somehow I can't get a seed on stm32f0discovery even thou I was back then (need to investigate tomorrow)
- anything else :-)?

PeterKietzmann commented:
Some thoughts from today:
- not working on STM boards was my personal fault -> forget it :-)
- I have the feeling that the random seed initialization costs a bit power (most likely due to initialized ADC on the iolab-m3 board).  See attached testbed outputs where I used the default example w & w/o `USEMODULE=random` and hit "reboot" three times during the first minute or so.

Without random module compiled:
![no_random](https://cloud.githubusercontent.com/assets/7765855/26574288/120efeb4-4521-11e7-8438-3b223ca385e9.png)
With random module compiled:
![with_random](https://cloud.githubusercontent.com/assets/7765855/26574287/120edb5a-4521-11e7-8188-4c35112937b7.png)

Why again don't we have something like `adc_off` in the API?

- It is tolerable but not pretty that the `UNSATISFIED FEATURES` warning occurs now in case no `hwrng` is present

PeterKietzmann commented:
Concerning the power consumption I wasn't aware of the concept described [here](https://github.com/RIOT-OS/RIOT/pull/6563/files#diff-01ef67be93732407f7aa76c827aba954R33). Even though it seems to be implemented for the stm32f1 I assume the problem lies somewhere in there

PeterKietzmann commented:
@haukepetersen please have a look at my latest changes in commit "test":
- separated the `random_seed` module from `random` so it is no dependency any more
- slightly adapted the `sc_random ` shell module to use `random_seed` instead of `xtimer` (in case the seed module was build)
- changed/added Makefile of the default example to use both `random` and `random_seed
- added a global variable to save the seed and additionally added two simple convenience functions to set and get the variable value
- adapted `auto_init` to initialize the random number generator with a random seed, in case the module was build (not so pretty but I see no other elegant way with this auto initialization concept)

PeterKietzmann commented:
@MichelRottleuthner you did some energy tests with this PR. Could you share some insights and the outcome, in a nutshell?!

MichelRottleuthner commented:
@PeterKietzmann I used the iolab-m3 board as you did in your Measurement. As I had no fitting connector at hand to measure the isolated MCU consumption I measured the total consumption of the board via USB. 
I measured with 100kS, default example with all LEDs turned off. 

Without random module:
![module_random_inactive_no_led_100ks](https://user-images.githubusercontent.com/20105554/29309282-54749a9c-81a9-11e7-9ed8-b33c1caafe24.png)

With `USEMODULE+=random` and `USEMODULE+=random_seed`:
![module_random_active_no_led_100ks](https://user-images.githubusercontent.com/20105554/29309276-4eadb1c0-81a9-11e7-8acd-72509b755f2c.png)

haukepetersen commented on Makefile.dep:
```diff
@@ -606,6 +606,14 @@ ifneq (,$(filter random,$(USEMODULE)))
+  USEMODULE += xtimer
+
+  FEATURES_REQUIRED += periph_hwrng
+  FEATURES_REQUIRED += periph_adc
```
Hmm, these dependencies still don't reflect whats actually done in the module. What should be reflected here is the actual usage and dependency of the modules used: 
`if (hwrng available) then (hwrng) else luid AND if (adc available) then adc else xtimer` (sorry, hard to read...).

haukepetersen replied:
if expressed like it is currently, there might be cases where e.g. `hwrng` is present but `xtimer` is included and initialized anyway (-> using power), even if `xtimer` is not used anywhere...

PeterKietzmann replied:
Correct and I also don't really like it. But IMO there is currently no other way. We do not know which resources are available at the time we pull in dependencies. So no way to conditionally include them. Am I mistaken?

kaspar030 replied:
Can you try matching against FEATURES_PROVIDED?

PeterKietzmann replied:
@kaspar030 `FEATURES_PROVIDED` is not set at that point in time. I re-checked it and we already had this problem in #6852. Am I blind to see your proposals? Can you give an example of what works?

haukepetersen replied:
What happens if you use `FEATURES_OPTIONAL` and then check in the implementation which ones are actually included?

haukepetersen replied:
But I think we desperately need to rethink the dependency system in the near future, this is getting annoying...

haukepetersen commented on examples/default/Makefile:
```diff
@@ -34,6 +34,12 @@ USEMODULE += ps
+# Using this module might print warnings about unsatisfied features such as HWRNG
+# and ADC, depending on your hardware resources. That must not necessarily mean
+# that the seed is bad. Have a look at the random_seed documentation.
+# USEMODULE += random_seed
```
This I really don't like. We must find a way so that the fitting/optimal modules for the seed module are included in a way that doesn't produce compile time warnings! I mean this is kind the idea for the fallback mechanism proposed here...

PeterKietzmann replied:
See above...

haukepetersen commented on sys/include/random_seed.h:
```diff
@@ -0,0 +1,92 @@
+/**
+ * @brief   Set default ADC line for random seed
+ *
+ * @todo    Define by board
```
I wouldn't call this a todo: this is a typical default value, so the more intuitive comment would be something like: `@note Override in your board configuration if needed`

haukepetersen commented on sys/include/random_seed.h:
```diff
@@ -0,0 +1,92 @@
+#endif
+
+/**
+ *  Global random seed value
```
missing `@brief`

haukepetersen commented on sys/include/random_seed.h:
```diff
@@ -0,0 +1,92 @@
+ *          2. A combination of single bits taken from multiple ADC samples and
+               the CPU ID, in case an, ADC is present.
+ *          3. A combination of the CPU iD and the system time value. Please note
+               that the CPU ID might equal between similar MCUs for some families.
```
I think this description needs to be more precise: which bytes are taken from which module and how many bits are taken from each ADC reading and so on...

haukepetersen commented on sys/include/random_seed.h:
```diff
@@ -0,0 +1,92 @@
+ *
+ * @return  the "unique" seed
+ */
+uint32_t random_prng_seed(void);
```
the function name does not fit our generic naming scheme: functions should start with the full module name as prefix. So `random_seed_generate()` would be IMHO the obvious choice here.

haukepetersen commented on sys/include/random_seed.h:
```diff
@@ -0,0 +1,92 @@
+
+ * @param[in] seed  The seed value to set
+ */
+void random_prng_set_global_seed(uint32_t seed);
```
I am not sure if these kind of getter/setter functions are really needed (as you also export the global variable holding the value...). But if you think we really need these functions, you should at least make them `static inline` and pull their implementation directly in the header.

haukepetersen commented on sys/random_seed/random_seed.c:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    /* use LSB of each ADC sample and shift 16
+        of these LSBs to form an 8-bit integer */
```
hmm, 16 shift operations to form a 8-bit integer? I think you mean 16-bit integer?

haukepetersen commented on sys/random_seed/random_seed.c:
```diff
@@ -0,0 +1,105 @@
+        val |= (sample & 0x01) << (iter);
+    }
+
+    /* TODO turn off ADC ??? */
```
nope: maybe also have a look at #6563

haukepetersen commented on sys/random_seed/random_seed.c:
```diff
@@ -0,0 +1,105 @@
+#else
+
+    /* Get three bytes ID. Upper byte will be filled by
+       different source */
```
the comment is not consistent with the 16-bit value we read from the ADC...

haukepetersen commented on sys/random_seed/random_seed.c:
```diff
@@ -0,0 +1,105 @@
+uint32_t random_prng_get_global_seed(void)
+{
+    return global_random_seed;
+}
```
newline missing

haukepetersen commented:
Some comments on the code above. 

I can't recall and try to remember: why did we need to save the global seed for later use again?

haukepetersen commented:
Oh, and regarding the power consumption: the usage of the ADC should in theory not lead to and additional energy being used. The driver is implemented (or at least should be) so that it powers down the ADC peripheral after each sample taken. So this needs to be investigated in detail why this is not the case.

I have an idea though what could be causing this: currently we enable/disable the peripheral clock each time we use the ADC. As this seems to be a good thing in theory, we have seen on some MCUs, that disabling the periph clock after using a peripheral does actually lead to way higher energy consumption that just leaving the peripheral clock enabled. So this is something we could look into.

PeterKietzmann commented:
@haukepetersen thanks for your review. There is IMO one blocker which needs further discussion (Makefiles dependencies commented above). The other comments I will address later. Concerning the global state we want to have a system-wide 'unique' number available, though I don't exactly have the use case in mind right now. I treated it as a 'requirement'.

PeterKietzmann commented:
@haukepetersen have a look at my latest commit where I've addressed your comments, except the one about build dependencies.

haukepetersen commented on sys/include/random_seed.h:
```diff
@@ -0,0 +1,102 @@
+ * @brief   This function returns the value of global_random_seed
+
+ * @return The global random seed value
+ * @return -1, if seed value is not set
```
not true -> `UINT32_MAX` if not set. But even this is tricky, as there is a 1 in a 4 billion chance, that this value is actually the generated seeed, is it not? So I would propose to remove this error case all together and add a note in the description, like e.g. `@note    Only use this function after you have set a global seed value`

PeterKietzmann commented:
@haukepetersen thanks againl,  I addressed your comments. Using `FEATURES_OPTIONAL` in the dependencies list "fixes" compile time warnings but does not fix the unconditional build of `xtimer` or `lluid` (by no surprise).

PeterKietzmann commented:
@kaspar030 what is your change request? Is it the the feature dependent include of modules or anything else?

kaspar030 commented:
I'm sorry to chip in late, but IMO, random number generators are to important to do half assed.

We've already had a couple of PRs trying to solve some of the problems this PR solves, and all of them have been NACK'ed, for good reason. This PR is more of the same.

How about we try, for once, to systematically find out what our requirements for RNGs are and how we can solve them?

I see the following problems with this PR...

1. this PR will leave the random function "quality" in total limbo state from the programmers perspective.
Writing code against "random_get()" might yield a random number from any of the following:

- unknown
- prng with default seed (image-unique)
- prng with seed from cpuid (node-unique?)
- prng with seed from periph/hwrng (hwrng unknown quality)
- prng with seed from xtimer (probably image-unique)
- prng with seed from ADC with selfmade implementation (implementation dependent random source with unkown quality)

- any combination of the above

2. "ADC random source" should be its own module

What I mean with this is that if we come up with a "use ADC as random source" (which is totally valid), why is that hidden in a "PRNG seed" module? It should be designed around a proper RNG interface, which has already been discussed e.g., on devel@.

3. ifdefs...

As there is no RNG interface that allows multiple *different* RNGs using the same interface, the whole module is a mess of "#ifdef/else/...". This is hard to review, hard to maintain, hard to extend. With a proper RNG API below, the whole module would, (apart from not needed anymore), by *much* shorter.
Also, with our current and growing hardware support in mind, I'm sure down the line noone will know which combination a particular board will be using. "compile and watch the output" will be the method of choice for finding out.

4. luid & ADC & xtimer combination seems arbitrary

"Get three bytes ID. Upper byte(s) will be filled by different source", 
"left shift by 16 to fill upper byte of uint32_t res one byte of the LUID will be overwritten",
"left shift by 24 to fill upper byte of uint32_t res"

Why each one, exactly?

---

This is what, off the top of my head, what we need in terms of random number generation:

- [ ] don't care about random quality
- [ ] good enough prevention network collisions for random-backoff protocols 
- [ ] usable for repeatable experiments ("predictable random numbers")
- [ ] properly seed suitable PRNG for crypto application
- [ ] directly use for strong crypto
- [ ] make sure code only runs on an RNG that suits its requirements

Without this PR, we could've checked the first box. With this PR, we can additionally check the second. That might look like an improvement, but if only that is provided, the module can be simplified down to just use the luid module to generate a node-unique seed.

PeterKietzmann commented:
@kaspar030, thanks for your detailed feedback. I appreciate it! Some of your concerns I understand, with others I don't fully agree. Let me comment some:
>random number generators are to important to do half assed

First of all, this module aims to provide **only** a random **seed** for PRNGs by leaving the choice of the actual random number generator open. Concening the RNG I think @tcschmidt had a future candidate in mind.

>How about we try, for once, to systematically find out what our requirements for RNGs are and how we can solve them?

Seems reasonable. Should we put it on the mailing list (again?) or/and open a pad later, to collect requirements? In any way, let's wait until next week to give others a chance to comment this PR or join this discussion.

> this PR will leave the random function "quality" in total limbo state from the programmers perspective.

I won't call it "limbo". It is on purpose to keep seeding and the actual random number generation separate. So you can either call `random_get()` with or without seeding the generator. In my opinion this is straight forward. However, some might think different here. I guess we need more opinions on that.
>...yield a random number from any of the following:

You mention valid situations here. However, I personally think feeding a generator with one explicit value, for example only the hwrng, or the CPUID can be done by hand. This module actually had in mind to "provide the best seed by not caring how". Conversely I think to abstract all these (simple) configurations by one API might bloat the module, for sparse added value.

>"ADC random source" should be its own module

Fine with me. I agree this module can be either improved in its implementation and pulled out of the seed module.
You mentioned a discussion on devel@. Would you mind sending a pointer to the archive?

>As there is no RNG interface that allows multiple different RNGs using the same interface, the whole module is a mess of "#ifdef/else/...". This is hard to review, hard to maintain, hard to extend. With a proper RNG API below, the whole module would, (apart from not needed anymore), by much shorter.

Maybe I'm failing to see your idea here but IMO you can ifdef (which is not so pretty, I agree), have some kind of registry (which I found too much complexity here) or make the implementation more hardware dependent.

>I'm sure down the line noone will know which combination a particular board will be using

Well, as said before. This was more or less the situation I focused. You don't exactly know your resources, but you don't care and get the "best" result.  And for those that are interested they can still have a look at boards features or enable `DEBUG` messages in the implementation.

>luid & ADC & xtimer combination seems arbitrary

I agree this seems a bit "random" ;-) but fortunately this could be adapted easily. Concerning the byte-wise combination I would refer to @tcschmidt who proposed it.

My comments to your wish list relating to my current implementation:

>- [x] don't care about random quality
> -[x] good enough prevention network collisions for random-backoff protocols 
>- [x] usable for repeatable experiments ("predictable random numbers")

Not a problem of this module IMO. Don't use a radom seed if you want prediction. Additionally the seer+RNG combination will easily give you random numbers to prevent for example network collisions.
>- [ ] properly seed suitable PRNG for crypto application
>- [ ] directly use for strong crypto

I don't dare to check these boxes even though such ADC mechanisms seem to be common in this area. So we are at least prepared.

- [x] make sure code only runs on an RNG that suits its requirements

I checked the box eveny though I see this module independent from the random number generator.

Now the question is, how do we proceed. To summarize the above:
- Collect requirements. Let's give it some more days and get to the mailing list again, probably open a pad?
- have a look at previous discussions on devel (@kaspar030 please provide a link)
- agree on the level of abstraction inside the module as well as to the user and the (co-) existence of a random seed module besides a random number generation module
- get others opinions?

haukepetersen commented:
First of all, we need to be clear on what we discuss here: do we simply discuss the creation of a seed value (as solely proposed in this PR), or do we want to go into the overall random number generation debate here?

I agree though, that we need to come up with an overall concept that can check during compile time (or at least during system initialization), if a certain 'quality level' of a random number is provided, which fits the applications needs. While reading the comments above, I was thinking of the following: we could introduce numerical 'quality' values like this:
```
1 - don't care about random quality
3 - PRNG good enough for network protocols
4 - PRNG good enough for standard crypto application
5 - high quality (P)RNG to use for strong crypto

--> having a constant seed for experiments is a special case...
```
An application could then assert a certain level (needs >= x). A seed generator as proposed in this PR could then for example figure out during compile time depending on the available modules, which level it can provide at the most...

So one very important question that we have to answer: do we ever see our systems running more than a single (P)RNG in parallel, so different seed values/mechanisms and also different (software)PRNGs at the same time? I am actually not so sure about this...

kaspar030 commented:
> You mentioned a discussion on devel@. Would you mind sending a pointer to the archive?

https://lists.riot-os.org/pipermail/devel/2017-March/005022.html (there's a rng API proposal somewhere in there)

> First of all, we need to be clear on what we discuss here: do we simply discuss the creation of a seed 
> value (as solely proposed in this PR), or do we want to go into the overall random number generation > debate here?

IMO, this goes hand in hand. No matter how much we shuffle around bits, generating a seed is done either from a) static (possibly node specific) data or b) from a random source. The latter should be hidden behind a proper interface, not hidden in a seed() function mixing a) and b) with unexplicable byte shuffling and then running it through the ifdef machinery of a >13k build matrix.

> So one very important question that we have to answer: do we ever see our systems running more 
 than a single (P)RNG in parallel, so different seed values/mechanisms and also different (software)PRNGs at the same time? I am actually not so sure about this...

high quality PRNGs are slow, which might make it advisable to use simpler ones in high speed low randomness applications. Otherwise the random backoff time calculation might reduce network throughput...

> It is on purpose to keep seeding and the actual random number generation separate.

How? Generating a seed is either generating a random number, or combining existing data. That's my main qualm about this PR, it mixes the concepts.

> This module actually had in mind to "provide the best seed by not caring how".

So from "no seed at all" we've generated a seed that ranges from "node-unique seeded" to "maybe hwrng seeded". That doesn't improve actual usability of the "seeded" PRNG for crypto applications. Worse, it makes it look like "board has adc or hwrng, seed will be fine".

It does provide a node-unique seed, but so does plain luid.

I'll create a thread on devel@...

PeterKietzmann commented:
Besides of all concerns it seems to me the first one we should clarify is whether we want a common random system which cares about seeds **and** random number generation or keep both features separate (independent of what I intended to provide with this PR).  
> high quality PRNGs are slow, which might make it advisable to use simpler ones in high speed low 

IMO this motivates to have both somehow separated for easier configuration. And that is the point where I started this PR: we have a couple of PRNGs, currently not properly seeded, and this gap I intend to fix with one module.

> Worse, it makes it look like "board has adc or hwrng, seed will be fine".

Well, from that perspective it wasn't too bad to have compile time warnings with `FEATURES_RPVODED` (see discussion above) :-)

However, I'm not completely against harmonizing seeding and number generation in one API but this requires a rework of the whole system. So let's get back to devel again ...

tcschmidt commented:
> First of all, we need to be clear on what we discuss here: do we simply discuss the creation of a seed value (as solely proposed in this PR), or do we want to go into the overall random number generation debate here?

I don't see so much of a debate here: random number generation just needs a good understanding of the field ... then it is very simple.
(There is a report from Berkely Lawrence Livermore Laboratory, though, that says: Never use an OS PRNG, because OS people just don't understand ;) ). 

> 
> I agree though, that we need to come up with an overall concept that can check during compile time (or at least during system initialization), if a certain 'quality level' of a random number is provided, which fits the applications needs. While reading the comments above, I was thinking of the following: we could introduce numerical 'quality' values like this:
> ```
> 1 - don't care about random quality
> 3 - PRNG good enough for network protocols
> 4 - PRNG good enough for standard crypto application
> 5 - high quality (P)RNG to use for strong crypto
> 

This sounds completely weird to me, who needs this? The OS just needs to provide a statistically decent, light-weight PRNG and a qualified seeding function for everyday purposes. 

If crypto-safety is needed, this must be done in parallel. Either by the application, or the OS makes an offering. But there is  no sense in offering anything like 1. or 4. , nor in doing statistical and crypto quality in the approach.
 
> --> having a constant seed for experiments is a special case...

The seed must be in any case accessible - that is needed by the application and on the side allows for reproducible runs.

> ```
> An application could then assert a certain level (needs >= x). A seed generator as proposed in this PR could then for example figure out during compile time depending on the available modules, which level it can provide at the most...

This is just overly complicated, who would need this?? 

> 
> So one very important question that we have to answer: do we ever see our systems running more than a single (P)RNG in parallel, so different seed values/mechanisms and also different (software)PRNGs at the same time? I am actually not so sure about this...

This question is simple to answer: It does not matter, if seed handling is accessible to the user.

The actual seeding question is:
  (1) Do we want a seed function only for auto_init?  Then - aside from crypto - it is enough to be unique.
  (2) Do we want a seed function for repeated use? I.e., different threads want different numbers by calling 'seed'? Then we need to take some observable state into account.

I would suggest to solve (1) for the moment. Applications have other sources of generating its seed. In particular, they could use the output of rand() (from auto_init) to seed their private random sequence.

tcschmidt commented:
@kaspar030:
 
> 
> ... with unexplicable byte shuffling and then running it through the ifdef machinery of a >13k build matrix

I agree that a unique, HW-independent generation of a unique seed on auto_init would be desirable. One way I learned about: Read uninitialized memory (but careful, there is some IPR on it). 

> 
> high quality PRNGs are slow, which might make it advisable to use simpler ones in high speed low randomness applications. 

This statement is not true - the execution complexity is not related to statistical randomness. A high-quality (much-much better than the complex Mersenne Twister ) at low(est)-complexity PRNG is https://en.wikipedia.org/wiki/Xorshift

Crypto-safety is a different thing - here we talk about 'unpredictability' (Entropy) instead of randomness.
 
> 
> > It is on purpose to keep seeding and the actual random number generation separate.
> 
> How? Generating a seed is either generating a random number, or combining existing data. That's my main qualm about this PR, it mixes the concepts.

??? I cannot understand this statement: Generating a seed is about generating a unique number. Generating random numbers is the problem of statistical sequences (not a single number). Generating an unpredictable number (for crypto) is about Entropy harvesting. 

Each of this can (and should be) solved separately and that's it.

> 
> > This module actually had in mind to "provide the best seed by not caring how".
> 
> So from "no seed at all" we've generated a seed that ranges from "node-unique seeded" to "maybe hwrng seeded". That doesn't improve actual usability of the "seeded" PRNG for crypto applications. Worse, it makes it look like "board has adc or hwrng, seed will be fine".
> 
> It does provide a node-unique seed, but so does plain luid.
>
The objective of the seed function is to provide a inter-node unique ID.
Locally unique IDs are not helpful here - actually we do need a (almost) globally unique ID to have random sequences on neighboring nodes run differently.
 
In any case: this endless discussion about a PRNG is a straw man's debate. Can we ASAP come to the following natural solution:

 * Function rand(<seed (optional input=0)>) -> implements the Xorshift
 * auto_init(seed) that assures rand has a unique starting point on each system?

In doing so, we are (a) clear, (b) appropriate, and (c) better than Linux.  I guess Peter will just provide a - never seen so good - solution.

kaspar030 commented:
> I don't see so much of a debate here: random number generation just needs a good understanding of the field ... then it is very simple.

The debate here is not about the generation of random numbers. It is about how to organize code, and how to provide an API that provides instead of misleads.

> This sounds completely weird to me, who needs this?

Maybe n levels of safety are a bit too much, but IMO it is very important to have a clear way of choosing an RNG by it's properties.

e.g.,

```
void cypto_create_key(uint8_t *out) {
    random_bytes(out, KEY_LEN);
}
```

vs

```
void cypto_create_key(uint8_t *out) {
    random_bytes_cryptosafe(out, KEY_LEN);
}
```

After all the discussion about how to seed PRNG's, which one might get compiled using unseeded tinymt32 without anyone realizing, and which one can be made to fail if no proper RNG is available?

kaspar030 commented:
> Function rand(<seed (optional input=0)>) -> implements the Xorshift

#7545 (xorshift not set as default, yet)

> auto_init(seed) that assures rand has a unique starting point on each system?

#7547 (for platforms providing cpuid, warns otherwise)

kaspar030 commented:
> Read uninitialized memory

Does this mean uninitialized memory after cold boot? Isn't that all zeros? If not, that would be very convenient.

tcschmidt commented:
> > I don't see so much of a debate here: random number generation just needs a good understanding of the field ... then it is very simple.
> 
> The debate here is not about the generation of random numbers. It is about how to organize code, and how to provide an API that provides instead of misleads.

D'accord: but the problem of organizing code is trivial. The standard use case of PRNGs is that you want to have a statistical pseudo-random sequence, which is unique (why you need a unique seed). That's it - just call rand(). But every programmer needs< an inevitable option of controlling the seed, that's why you need rand(<seed>).

> 
> > This sounds completely weird to me, who needs this?
> 
> Maybe n levels of safety are a bit too much, but IMO it is very important to have a clear way of choosing an RNG by it's properties.

Safety is the wrong category here: choose the algorithms wisely and make them transparent. That*s it. Normally, you want to be statistically good (means represent your distribution in high, independent dimensions) . 
If we offer a crypto-Version, do the same: choose state of the art solution and make it transparent. 

That's it. If someone wants something else, he/she needs to implement themselves.

> 
> e.g.,
> 
> ```
> void cypto_create_key(uint8_t *out) {
>     random_bytes(out, KEY_LEN);
> }
> ```
> 
> vs
> 
> ```
> void cypto_create_key(uint8_t *out) {
>     random_bytes_cryptosafe(out, KEY_LEN);
> }
> 

Don't see the seeding difference. Normally you just go for 
 ```
 uint* rand(uint*:seed ) {
    randombytes_calculated_from(seed);
 }

> 
> After all the discussion about how to seed PRNG's, which one might get compiled using unseeded tinymt32 without anyone realizing, and which one can be made to fail if no proper RNG is available?

??? It should be really easy to resolve this issue.

tcschmidt commented:
> > Read uninitialized memory
> 
> Does this mean uninitialized memory after cold boot? Isn't that all zeros? If not, that would be very convenient.

I just learned from Pekka - I have no experience with that. Should be experimentally evaluated.

PeterKietzmann commented:
Thanks for productive discussions. I will close this PR with memo label set. For now we go with a simpler and cleaner solution with only board-unique seeds as proposed in #7547
